### PR TITLE
feat: robotics case submission notification (FPLA-720)

### PR DESCRIPTION
**Pull request by Gogetter**
### JIRA link (if applicable) ###
[FPLA-720](https://tools.hmcts.net/jira/browse/FPLA-720)

### Change description ###
Email Notification to Robotics on Case Submissions (with json file attachment).

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        RoboticsData preparedRoboticsData = roboticsDataService.prepareRoboticsData(caseData);
+
+        assertThat(preparedRoboticsData.getApplicationType()).isEqualTo(
+            "Care order,Education supervision order,Emergency protection order,"
+                + "Other order under part 4 of the Children Act 1989");
+    }
+
+    @Test
+    void shouldReturnExpectedRoboticsDataJsonString() throws IOException {
+        String expectedRoboticsDataJson = objectMapper.writeValueAsString(expectedRoboticsData(
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java]:
```suggestion
        String expectedRoboticsJson = objectMapper.writeValueAsString(expectedRoboticsData(
```

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            SUPERVISION_ORDER.getLabel()));
+
+        CaseData caseData = prepareCaseDataWithOrderType(INTERIM_SUPERVISION_ORDER);
+
+        RoboticsData roboticsData = roboticsDataService.prepareRoboticsData(caseData);
+        String returnedRoboticsDataJson = roboticsDataService.convertRoboticsDataToJson(roboticsData);
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java]:
```suggestion
        String returnedRoboticsJson = roboticsDataService.convertRoboticsDataToJson(roboticsData);
```

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+        assertEquals(expectedRoboticsDataJson, returnedRoboticsDataJson, true);
+    }
+
+    @Test
+    void shouldReturnRoboticsJsonDataWithoutSolicitorNode() throws IOException {
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java]: when

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+        Map<String, Object> roboticsDataMap = objectMapper.reader()
+            .forType(new TypeReference<Map<String, Object>>() {})
+            .readValue(returnedRoboticsDataJson);
+
+        assertThat(roboticsDataMap).doesNotContainKey("solicitor");
+    }
+
+    @Test
+    void shouldReturnRoboticsDataWithNullAllocation() throws IOException {
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java]: when

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            .allocationProposal(null)
+            .build();
+
+        RoboticsData roboticsData = roboticsDataService.prepareRoboticsData(caseData);
+
+        assertThat(roboticsData.getAllocation()).isNull();
+    }
+
+    @Test
+    void shouldReturnRoboticsDataWithExpectedlAllocation() throws IOException {
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java]: when

File: service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                .build())
+            .build();
+
+        RoboticsData roboticsData = roboticsDataService.prepareRoboticsData(caseData);
+
+        assertThat(roboticsData.getAllocation()).isEqualTo(expectedAllocation);
+    }
+
+    @Test
+    void shouldReturnRoboticsDataWithEducationSupervisionOrderLabelValueAsApplicationType() throws IOException {
```
Comment (swalker125) [inline service/src/test/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataServiceTest.java]: when

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/AddCaseNumberController.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    private final ObjectMapper objectMapper;
+    private final ApplicationEventPublisher applicationEventPublisher;
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = objectMapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        if (!isAlphanumeric(caseData.getFamilyManCaseNumber())) {
+            caseDetails.getData().keySet().removeIf(key -> key.equalsIgnoreCase("familyManCaseNumber"));
```
[the remark this answers is not in the document]
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/AddCaseNumberController.java]: good question but I remember `familyManCaseNumber`, was being persisted even when condition failed. But will recheck again

File: service/src/main/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataService.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static org.apache.commons.lang3.math.NumberUtils.toInt;
+import static uk.gov.hmcts.reform.fpl.enums.OrderType.CARE_ORDER;
+import static uk.gov.hmcts.reform.fpl.enums.OrderType.EDUCATION_SUPERVISION_ORDER;
+import static uk.gov.hmcts.reform.fpl.enums.OrderType.EMERGENCY_PROTECTION_ORDER;
+import static uk.gov.hmcts.reform.fpl.enums.OrderType.OTHER;
+import static uk.gov.hmcts.reform.fpl.enums.OrderType.SUPERVISION_ORDER;
+import static uk.gov.hmcts.reform.fpl.model.robotics.Gender.convertStringToGender;
+
+@Slf4j
```
[the remark this answers is not in the document]
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/service/robotics/RoboticsDataService.java]: it's used in `convertRoboticsDataToJson(...)` method

File: service/src/main/java/uk/gov/hmcts/reform/fpl/model/email/EmailData.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import lombok.Value;
+
+import java.util.Set;
+
+import static org.springframework.util.CollectionUtils.isEmpty;
+
+@Value
+@Builder(toBuilder = true)
+public class EmailData {
+    private final String to;
```
[the remark this answers is not in the document]
Comment (Gogetter) [reply service/src/main/java/uk/gov/hmcts/reform/fpl/model/email/EmailData.java]: to/recipient works for me. Updated

File: charts/fpl-case-service/values.preview.template.yaml [hunk 10]
```diff
@@ -15,6 +15,11 @@ java:
     SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_ISSUERURI: https://forgerock-am.service.core-compute-idam-aat.internal:8443/openam/oauth2/hmcts
     SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWKSETURI: https://idam-api.aat.platform.hmcts.net/o/jwks
     GATEWAY_URL: https://gateway-${SERVICE_FQDN}
+    SPRING_MAIL_HOST: mta.reform.hmcts.net
+    SPRING_MAIL_PORT: 25
+    SPRING_MAIL_TEST_CONNECTION: false
+    SPRING_MAIL_PROPERTIES_MAIL_SMTP_STARTTLS_ENABLE: true
+    SPRING_MAIL_PROPERTIES_MAIL_SMTP_SSL_TRUST: mta.reform.hmcts.net
```
Comment (akomisarek) [inline charts/fpl-case-service/values.preview.template.yaml]: As we are using the same settings, maybe just define them in values file?

File: charts/fpl-case-service/values.aat.template.yaml [hunk 11]
```diff
@@ -16,6 +16,11 @@ java:
     SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_ISSUERURI: https://forgerock-am.service.core-compute-idam-aat.internal:8443/openam/oauth2/hmcts
     SPRING_SECURITY_OAUTH2_RESOURCESERVER_JWT_JWKSETURI: https://idam-api.aat.platform.hmcts.net/o/jwks
     GATEWAY_URL: http://gateway-ccd.aat.platform.hmcts.net
+    SPRING_MAIL_HOST: mta.reform.hmcts.net
+    SPRING_MAIL_PORT: 25
+    SPRING_MAIL_TEST_CONNECTION: false
+    SPRING_MAIL_PROPERTIES_MAIL_SMTP_STARTTLS_ENABLE: true
+    SPRING_MAIL_PROPERTIES_MAIL_SMTP_SSL_TRUST: mta.reform.hmcts.net
```
Comment (akomisarek) [inline charts/fpl-case-service/values.aat.template.yaml]: Hmm I am not sure if host shouldn't be a secret. I don't think giving some of our internal hosts is good thing.

File: charts/fpl-case-service/values.aat.template.yaml [hunk 12]
```diff
@@ -34,4 +39,6 @@ java:
         - local-authority-fallback-inbox
         - system-update-user-username
         - system-update-user-password
+        - robotics-notification-sender
+        - robotics-notification-recipient
```
Comment (akomisarek) [inline charts/fpl-case-service/values.aat.template.yaml]: Please update fpl-case-service-configuratin (repo) with new secrets, I forgot about that step in some of other prs!

File: service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/AddCaseNumberController.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .errors(validationErrors(caseDetails))
+            .build();
+    }
+
+    @PostMapping("/submitted")
+    public void handleSubmittedEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseData caseData = mapper.convertValue(callbackRequest.getCaseDetails().getData(), CaseData.class);
+        applicationEventPublisher.publishEvent(new CaseSubmittedEvent(caseData));
```
Comment (akomisarek) [inline service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/AddCaseNumberController.java]: Event name is wrong - we haven't submitted case here.

File: charts/fpl-case-service/values.aat.template.yaml (hunk 11 above)
Comment (Gogetter) [reply charts/fpl-case-service/values.aat.template.yaml]: yes wasn't sure about the host. Think we should add as a secret? @akomisarek

File: charts/fpl-case-service/values.preview.template.yaml (hunk 10 above)
Comment (Gogetter) [reply charts/fpl-case-service/values.preview.template.yaml]: yes, we could. updated

File: charts/fpl-case-service/values.aat.template.yaml (hunk 11 above)
Comment (akomisarek) [reply charts/fpl-case-service/values.aat.template.yaml]: I thought about this no secret necessary in my opinion.

**Pull request closed without merging**